osquery/database/database.cpp

```diff
@@ -366,16 +366,31 @@ Status ptreeToRapidJSON(const std::string& in, std::string& out) {
     return Status(1, "Failed to parse JSON");
   }
 
+  if (tree.empty()) {
+    JSON().toString(out);
+    return Status();
+  }
+
   auto json = JSON::newArray();
   for (const auto& t : tree) {
-    std::stringstream ss;
-    pt::write_json(ss, t.second);
-
-    rj::Document row;
-    if (row.Parse(ss.str()).HasParseError()) {
-      return Status(1, "Failed to serialize JSON");
+    JSON obj = json.newObject();
+
+    if (!t.second.empty()) {
+      // if the node is a subtree, we parse this directly with RJ
+      std::stringstream ss;
+      pt::write_json(ss, t.second);
+      if (obj.doc().Parse(ss.str()).HasParseError()) {
```

Comment:
Isn't parsing ptree generated json with rapidjson dangerous? I think we need to parse everything manually, maybe recursively.

Reply:
Yeah see the above comment. This function wasn't intended to be a perfect conversion, it was just meant to handle the upgrade cases to get us past that error we were seeing with the object not being a list. I think we might be able to get away with more limited functionality on this function, as it was mostly designed to convert the query data contents, which ended up being something that RJ could read in from ptree written out as a json string.